File: web/translate-file.pl
#!/usr/bin/perl

# Puxcobot - A robot to play Coup in Esperanto (Puĉo)
# Copyright (C) 2011, 2020  Neil Roberts
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use Getopt::Std;

sub process_arguments
{
    my %opts = ();
    my $ret;

    $ret = getopts("l:i:o:", \%opts);

    if (!$ret ||
        !defined($opts{"l"}) ||
        !defined($opts{"i"}) ||
        !defined($opts{"o"}) ||
        @ARGV > 0)
    {
        print STDERR "usage: $0 -l <translation-file> " .
            "-i <input-file> -o <output-file>\n";
        exit 1;
    }

    return %opts;
}

sub cleanup_value
{
    my ($value) = @_;

    $value =~ s/^\s+//;
    $value =~ s/\s+$//;
    $value =~ s/\s+/ /g;

    return $value;
}

sub read_translation_file
{
    my ($translation_filename) = @_;
    my %trans;
    my $key;
    my $value = "";

    open(my $infile, '<', $translation_filename)
        or die("Couldn't open $translation_filename");

    while (my $line = <$infile>)
    {
        if ($line =~ /^@([^@]+)\@$/)
        {
            if (defined($key))
            {
                $trans{$key} = cleanup_value($value);
                $value = "";
            }

            $key = $1;
        }
        elsif (defined($key))
        {
            $value .= $line;
        }
    }

    if (defined($key))
    {
        $trans{$key} = cleanup_value($value);
    }

    close($infile);

    return %trans;
}

sub make_regexp
{
    return "@(" . join("|", map(quotemeta($_), @_)) . "|SELECTED_[A-Z]+)@";
}

sub get_replacement
{
    my ($trans, $key) = @_;

    if ($key =~ /\ASELECTED_([A-Z]+)\z/)
    {
        if ($1 eq uc($trans->{"LANG_CODE"}))
        {
            return "selected=\"selected\"";
        }
        else
        {
            return "";
        }
    }
    else
    {
        return $trans->{$key};
    }
}

sub translate_file
{
    my ($filename, $trans, $infile, $outfile) = @_;
    my $re = make_regexp(keys(%$trans));
    my $line_num = 1;

    while (my $line = <$infile>)
    {
        $line =~ s/$re/get_replacement($trans, $1)/eg;
        print $outfile $line;

        my @bad_replacements = $line =~ /@([a-z_0-9A-Z]+)@/g;

        for my $br (@bad_replacements)
        {
            print STDERR "warning: $filename:$line_num: " .
                "untranslated key: $br\n";
        }

        $line_num++;
    }
}

my %opts = process_arguments();
my %trans = read_translation_file($opts{"l"});

open(my $infile, "<", $opts{'i'})
    or die("Couldn't open $opts{'i'}");
open(my $outfile, ">", $opts{'o'})
    or die("Couldn't open $opts{'o'}");

translate_file($opts{"i"}, \%trans, $infile, $outfile);

close($infile);
close($outfile);

File: web/translate-file.py
#!/usr/bin/python3

# Pucxobot - A bot and website to play some card games
# Copyright (C) 2011, 2020  Neil Roberts
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import re
import sys


def process_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument("-l", "--language",
                        help="Load translations. Can be specified multiple "
                        "times",
                        required=True,
                        metavar="translation-file",
                        action='append')
    parser.add_argument("-i", "--input",
                        help="Set the input file. Can be specified multiple "
                        "times",
                        required=True,
                        metavar="input-file",
                        action='append')
    parser.add_argument("-o", "--output",
                        help="Set the output file",
                        required=True,
                        metavar="output-file")
    return parser.parse_args()


def read_translation_file(filename):
    key_re = re.compile(r'^@([^@]+)@$')
    space_re = re.compile(r'[ \n]+') # don’t use \s to avoid replacing nbsp
    trans = {}
    key = None
    value = []

    def flush_value():
        if key is None:
            return

        trans[key] = space_re.sub(" ", "".join(value)).strip()
        value.clear()

    with open(filename, "rt", encoding="utf-8") as f:
        for line in f:
            md = key_re.match(line)

            if md is not None:
                flush_value()

                key = md.group(1)
            else:
                value.append(line)

    flush_value()

    return trans


def translate_files(trans, filenames_in, filename_out):
    key_re = re.compile(r'@(\w+)@')

    with open(filename_out, "wt", encoding="utf-8") as outfile:
        for filename_in in filenames_in:
            line_num = 1

            def get_replacement(md):
                k = md.group(1)
                try:
                    return trans[k]
                except KeyError:
                    print("warning: {}:{}: untranslated key: {}".
                          format(filename_in,
                                 line_num,
                                 k),
                          file=sys.stderr)
                    return md.group(0)

            with open(filename_in, "rt", encoding="utf-8") as infile:
                for line in infile:
                    print(key_re.sub(get_replacement, line),
                          file=outfile,
                          end='')

                line_num += 1


if __name__ == '__main__':
    args = process_arguments()

    trans = {}

    for translation_file in args.language:
        trans.update(read_translation_file(translation_file))

    translate_files(trans, args.input, args.output)
